### PR TITLE
Add SO_BROADCAST option to UDP sockets

### PR DESCRIPTION
**Pull request by ObsidianX**
**Subsystem**
ktor-network

**Motivation**
ktor-network provides a coroutine-compatible implementation of Sockets and Channels but is pretty bare-bones at the moment.  Broadcast UDP sockets are useful when advertising or discovering services on a network.  Without setting `SO_BROADCAST` on the socket, broadcast messages are not properly sent.

**Solution**
Add a `broadcast` option to `UDPSocketOptions` to configure a UDP socket in broadcast mode.

**Example**
```kotlin
val socket = aSocket(ActorSelectorManager(Dispatchers.IO))
  .udp()
  .bind(InetSocketAddress("192.168.0.100", 0)) {
    reuseAddress = true
    broadcast = true
  }
socket.send(
  Datagram(
    ByteReadPacket(ByteBuffer()),
    InetSocketAddress("192.168.0.255", 1234)
  )
)
```

### CODE REVIEW

File: ktor-network/jvm/src/io/ktor/network/sockets/JavaSocketOptions.kt [hunk 1]
```diff
@@ -9,6 +9,7 @@ import java.lang.reflect.*
 import java.nio.channels.*
 
 private const val SO_REUSEPORT = "SO_REUSEPORT"
+private const val SO_BROADCAST = "SO_BROADCAST"
```
Comment (cy6erGn0m) [inline ktor-network/jvm/src/io/ktor/network/sockets/JavaSocketOptions.kt]: this is unused

File: ktor-network/jvm/src/io/ktor/network/sockets/SocketOptions.kt [hunk 2]
```diff
@@ -117,6 +117,18 @@ sealed class SocketOptions(
      */
     class UDPSocketOptions internal constructor(customOptions: MutableMap<Any, Any?>) :
         PeerSocketOptions(customOptions) {
+        /**
+         * SO_BROADCAST socket option
+         */
+        var broadcast: Boolean = false
```
Comment (cy6erGn0m) [inline ktor-network/jvm/src/io/ktor/network/sockets/SocketOptions.kt]: are all udp kinds sockets can have this option set?

Comment (ObsidianX) [reply ktor-network/jvm/src/io/ktor/network/sockets/SocketOptions.kt]: Yes, SO_BROADCAST is strictly a datagram socket option.  You cannot send a message to the broadcast address without enabling it first.

File: ktor-network/jvm/src/io/ktor/network/sockets/JavaSocketOptions.kt (hunk 1 above)
Comment (ObsidianX) [reply ktor-network/jvm/src/io/ktor/network/sockets/JavaSocketOptions.kt]: Removed

File: ktor-network/jvm/src/io/ktor/network/sockets/SocketOptions.kt (hunk 2 above)
Comment (cy6erGn0m) [reply ktor-network/jvm/src/io/ktor/network/sockets/SocketOptions.kt]: What about "bound" and "connected" udps?

**Comment by Thomas-Vos** [comment]
@cy6erGn0m could this PR be merged, please? This is a really useful feature.

**Comment by Thomas-Vos** [comment]
Thanks for reopening the pull request, @e5l. What can I do to help? I tried this pull request on the master branch and it is working great in my project. Now I can send broadcasts on Android which didn't work before. Here is a small example how I am using this feature:

```kt
val socket = aSocket(createSelectorManager(dispatchers))
    .udp()
    .bind(NetworkAddress("0.0.0.0", 56700)) {
        reuseAddress = true
        broadcast = true
    }

// Somewhere else in the code
val address = NetworkAddress("255.255.255.255", 56700)
socket.send(
    Datagram(
        packet = createPacket(),
        address = address
    )
)

// Also still possible to send packets to a single device
val address2 = NetworkAddress("192.168.86.49", 56700)
socket.send(
    Datagram(
        packet = createPacket(),
        address = address2
    )
)
```

I would like to use this on iOS as well, but still need to look into that so I am not sure if there are any changes needed for broadcasts on iOS.

I am using the UDP sockets in Ktor to control LIFX lights (smart WiFi lights, UDP control over local network). This feature is really important to me, because I need it to send a "discover" broadcast on the local network to get the IP addresses of all available LIFX lights. If there is anything I can do to help, please let me know.

File: ktor-network/jvm/src/io/ktor/network/sockets/SocketOptions.kt (hunk 2 above)
Comment (Thomas-Vos) [reply ktor-network/jvm/src/io/ktor/network/sockets/SocketOptions.kt:123]: @cy6erGn0m what exactly would you like to know? I left a comment on this PR with my use case: https://github.com/ktorio/ktor/pull/1258#issuecomment-689026286.

**Comment by e5l** [comment]
Obsolete by #2075

**Pull request closed without merging**